core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -91,29 +98,38 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
             return;
         }
 
-        catalogsUpdateLock.lock();
+        Lock catalogLoadLock = catalogsUpdateLock.readLock();
+        catalogLoadLock.lock();
         try {
             if (stopped) {
                 return;
             }
 
-            for (CatalogProperties catalog : getMissingCatalogs(expectedCatalogs)) {
-                checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
-                CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
-                catalogs.put(catalog.getCatalogHandle(), newCatalog);
-                log.debug("Added catalog: " + catalog.getCatalogHandle());
-            }
+            CompletableFuture.allOf(
+                getMissingCatalogs(expectedCatalogs).stream()
+                        .map(catalog ->
+                                CompletableFuture.runAsync(() -> {
+                                    checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
+                                    catalogs.computeIfAbsent(catalog.getCatalogHandle(), ignore -> {
+                                        CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
+                                        log.debug("Added catalog: " + catalog.getCatalogHandle());
+                                        return newCatalog;
+                                    });
+                                }, executor))
+                        .collect(toImmutableList())
+                        .toArray(CompletableFuture[]::new)).join();
         }
         finally {
-            catalogsUpdateLock.unlock();
+            catalogLoadLock.unlock();
         }
     }
 
     @Override
     public void pruneCatalogs(Set<CatalogHandle> catalogsInUse)
     {
         List<CatalogConnector> removedCatalogs = new ArrayList<>();
-        catalogsUpdateLock.lock();
+        Lock catalogCleanUpLock = catalogsUpdateLock.writeLock();
```

Comment:
I think it would be cleaner to hoist the write and read locks to a field in the class. It will make it more obvious that different methods that are getting the write lock are getting and operating on the same lock.